graylog2-server/src/main/java/org/graylog2/shared/rest/resources/ProxiedResource.java

```diff
@@ -112,4 +114,82 @@ protected ProxiedResource(@Context HttpHeaders httpHeaders,
             }
         };
     }
+
+    /**
+     * Execute the given remote interface function on the master node.
+     * <p>
+     * This is used to forward an API request to the master node. It is useful in situations where an API call can
+     * only be executed on the master node.
+     * <p>
+     * The returned {@link MasterResponse} object is constructed from the remote response's status code and body.
+     */
+    protected <RemoteInterfaceType, RemoteCallResponseType> MasterResponse<RemoteCallResponseType> requestOnMaster(
+            Function<RemoteInterfaceType, Call<RemoteCallResponseType>> remoteInterfaceFunction,
+            Function<String, Optional<RemoteInterfaceType>> remoteInterfaceProvider
+    ) throws IOException {
+        final Node masterNode = nodeService.allActive().values().stream()
+                .filter(Node::isMaster)
+                .findFirst()
+                .orElseThrow(() -> new IllegalStateException("No active master node found"));
+
+        final RemoteInterfaceType remoteInterfaceType = remoteInterfaceProvider.apply(masterNode.getNodeId())
+                .orElseThrow(() -> new IllegalStateException("Master node " + masterNode.getNodeId() + " not found"));
+
+        final Call<RemoteCallResponseType> call = remoteInterfaceFunction.apply(remoteInterfaceType);
+        final Response<RemoteCallResponseType> response = call.execute();
+
+        final byte[] errorBody = response.errorBody() == null ? null : response.errorBody().bytes();
+
+        return MasterResponse.create(response.isSuccessful(), response.code(), response.body(), errorBody);
+    }
+
+    @AutoValue
+    protected static abstract class MasterResponse<ResponseType> {
+        /**
+         * Indicates whether the request has been successful or not.
+         * @return {@code true} for a successful request, {@code false} otherwise
+         */
+        public abstract boolean isSuccess();
+
+        /**
+         * Returns the HTTP status code of the response.
+         *
+         * @return HTTP status code
+         */
+        public abstract int code();
+
+        /**
+         * Returns the typed response object if the request was successful. Otherwise it returns {@code null}.
+         *
+         * @return typed response object or {@code null}
+         */
+        @Nullable
```

Comment:
Instead of making it `Nullable`, we should use an `Optional` here.

Reply:
:heavy_check_mark: